### ISSUE TITLE
New listener API that gets real suite/test objects as arguments and can modify them

### ISSUE THREAD

**Issue by spooning** [robotframework/robotframework#1208]
> _Originally submitted to [Google Code](http://code.google.com/p/robotframework/issues/detail?id=1208) by bryan.oakley on 27 Aug 2012_

I would like to see a listener mode that passes the actual objects to the listeners rather than just a few attributes. For example, I would like to be able to write a listener like this:

```
ROBOT_LISTENER_API_VERSION = 3
def start_suite(suite):
    print "starting suite", suite.name
```

Likewise for test cases and keywords.

Implementation is pretty straight-forward. In robot/output/listeners.py a new clause would be added for each method. For example:

```
def start_suite(self, suite):
    for li in self._listeners:
        if li.version == 1:
            li.call_method(li.start_suite, suite.name, suite.doc)
        elif li.version == 2:
            attrs = self._get_start_attrs(suite, 'metadata')
            attrs.update({'tests' : [t.name for t in suite.tests],
                          'suites': [s.name for s in suite.suites],
                          'totaltests': suite.get_test_count()})
            li.call_method(li.start_suite, suite.name, attrs)
        elif li.version == 3: 
            li.call_method(li.start_suite, suite)
```

This takes a big step forward toward being able to control a test run via the listener interface. The next step would be to put methods on the objects that control the state of the test run. For example, a start_suite listener might do something like 'suite.fail("blah blah")' to cause the suite to fail at the point the listener was called. 

The modification of the objects is not the subject of this feature request; that will come in a later feature request. This request is solely for the ability for the listener interface to pass the objects to the listeners. 

---

**UPDATE:** Jump to [this comment](https://github.com/robotframework/robotframework/issues/1208#issuecomment-164910769) to skip old discussion and to see how this new API is actually going to be implemented.

**Comment by aaltat** [comment]
Is there any plans this to make in 2.9?

**Comment by pekkaklarck** [comment]
This isn't currently in RF 2.9 scope, but can definitely be included if someone is interested to provide a pull request.

**Comment by pekkaklarck** [comment]
We try to get this into RF 2.9.1 already.

**Comment by vkruoso** [comment]
This is probably the wrong place to ask, but I was wondering why there is no `id` attribute about keywords (`start_keyword` and `end_keyword`). It seems explicitly removed as per [this code](https://github.com/robotframework/robotframework/blob/e3c22e382bd433ef539657d1ffc475704a700ecd/src/robot/output/listeners.py#L55-L56).

I'm assuming that when the interface provides the object itself we would have access to it. Can you explain why the `id` field has been excluded in the version 2 of the API for the keywords (95779ed8d5b5203ff0e52723a88b56f83cfc72dc)? Would you be open to a PR that adds it (it would make my life easier)?

**Comment by pekkaklarck** [comment]
This issue requires a bit too much changes to be safe in 2.9.x releases. Re-scoping back to 3.0.

@vkruoso: I don't remember why id wasn't added to keyword. Most likely because we only needed it with suites and tests. Could have been also that we were afraid that getting it causes performance issues (ids are always re-calculated) but I doubt that would really be an issue. In other words, I don't see why it couldn't be added to the v2 API. Please submit a separate issue about that.

**Comment by pekkaklarck** [comment]
OK, I'm going to look at this issue next. After the recent refactoring/rewrite of the old listener API, implementing the new API ought to be pretty easy. @jussimalinen and I in fact already prototyped that last week, and were able to get something done with a minimal effort. It could be that testing and documenting this could be more work than actual implementation.

Based on our prototype, the most important methods of the new API are going to be these:

```python
def start_suite(data, result):
    pass
def end_suite(data, result):
    pass
def start_test(data, result):
    pass
def end_test(data, result):
    pass
```

All of the above methods get two arguments, the model object of the currently executed test/suite and the corresponding result object. Both objects can be inspected freely and also modified in many ways. Modifying the current data object won't have any effect on the end results, though, because the result object is already created by it. Adding more tests to the suite or keywords to the test is possible, though. If the result object is modified, changes are seen also in the final results.

Notice that we do not plan to add `start_keyword` and `end_keyword` methods in RF 3.0. The main reasons are the following:
- There isn't enough context information available in the runners executing keywords to create model objects with information about the parent objects. This could obviously be fixed, and after the recent refactoring/rewrite of keyword execution that shouldn't even be too hard, but we just don't have time for that in RF 3.0.
- Modifying the keyword result object wouldn't automatically modify the resulting test. If, for example, keyword status would be turned to `FAIL`, Robot wouldn't automatically notice that and would continue executing further keywords and mark the test passed at the end. Also this issue could probably be resolved, but we, again, don't have enough time now.
- Most important use case for the new listener API can be accomplished with the suite/test methods.
- When we add no keyword related methods, we can design them freely, without taking backwards compatibility into account, later when we have more experience from the API.

The new API will most likely have also the following methods:

```python
def log_message(message):
    pass
def message(message):
    pass
```

These methods will behave like methods with same name in the old API, but they will get real `Message` objects as arguments instead of a dictionary. Modifying the message will mean that the logged message is changed.

Finally, the API is planned to have exactly same `xxx_file`, `xxx_import` and `close` methods as the old API.

Comments about the above plans are highly appreciated! They should be given soon because I'll start implementing this tomorrow.

**Comment by vkosuri** [comment]
Any possibility adding keywords to this listener?

```python
def start_keyword(data, result):
    pass
def end_keyword(data, result):
    pass
```

**Comment by pekkaklarck** [comment]
Hopefully yes but certainly not now. See the prevous comment for details.

**Comment by vkruoso** [comment]
I think this looks really good. I understand that there is no time to implement the keyword methods and they can be out for now as most use cases can live with that. Is your intention to deprecate the v2 API? If it is, I think it should be made only after the keywords methods are available.

**Comment by pekkaklarck** [comment]
No plan to deprecate API v2. Would break too many things, including RIDE, and it also works with Java.

**Comment by aaltat** [comment]
Will there be version 3?

**Comment by pekkaklarck** [comment]
Yes, the new API requires `ROBOT_LISTENER_API_VERSION = 3` attribute.

**Comment by pekkaklarck** [comment]
As commented in b121a4370bfd87d0ed9cc2832337ab0d95788741, implementation ought to be done. First tests have also been committed in 54051e5c4f273ac8eb4c5123226b0b1e62336aa8 and they can be used to see what kind of cool/weird things this new API supports. 

TODO:
- [ ] Test `log_message` and `message` methods
- [ ] Test `xxx_file`, `xxx_import` and `close` methods (work same way as in API v2)
- [ ] Test with library listeners
- [ ] Documentation

We'll release RF 3.0 beta 1 once this issue is ready. Early adopter feedback is appreciated already now!

**Comment by pekkaklarck** [comment]
We decided not to include `library_import`, `resource_import` and `variables_import` methods to listener v3 at this point. The main reason is that we would like to pass actual model objects representing libraries, resources and variable files to these methods so that they could be modified, but that is too much work for RF 3.0. The current model objects aren't good enough for this purpose and pretty big changes would be required also for import logic.

**Comment by pekkaklarck** [comment]
This new functionality is finally documented. Can you @jussimalinen review the docs? Comments from others would be very highly appreciated too, but unfortunately generated docs aren't available anywhere until beta 1 is released.

Once someone else has looked at the docs this issue can be considered done and beta 1 created. I will briefly look at `start/end_keyword` methods after the beta, but I'm still afraid they need to wait for 3.1 or 3.0.1.

**Comment by jussimalinen** [comment]
Read through the docs on iphone. Wrote a couple of line comments because modification on phone did not seem too fun. Looks good I would say.

**Comment by pekkaklarck** [comment]
Beta 1 is out and the new listener version documentation visible here:
http://robotframework.org/robotframework/3.0b1/RobotFrameworkUserGuide.html#listener-interface

Feedback welcome!

**Comment by aaltat** [comment]
Perhaps it's more an API documentation issue but would it be possible to list all object atributes where user can influence from listener interface. Or link to corresponding unit/acceptance test, to provide examples.

I didn't quite understand, why the first argument was called data. In my mind it didn't describe what the argument receives. Would runner or model describe better the argument. Or use suite and test as argument names in corresponding situation.

**Comment by pekkaklarck** [comment]
Thanks for the feedback @aaltat.

It's definitely more of a API documentation issue to explain attributes of the various model objects. It would probably be best to handle that as a separate issue listing attributes and other topics that needs documentation or that have documentation that could be enhanced. Even better if such an issue would be backed by a pull request.

Related to the new listener API, I think it is best to document what you cannot change. The biggest limitations is that changing the current data model doesn't change results because the result model has already been created (you should update result instead) and that trying to change results when related information is already written to output.xml doesn't work either. These both are documented in the UG, but I'm open for ideas how to enhance the documentation.

I was initially thinking about adding a link to the acceptance tests from the UG, but in the end I thoght the included examples already covered everything and were much more simpler. If you think they would be useful, I can consider adding them, but we can also consider adding more examples. Tests and listener that are used for testing this functionality are here:
https://github.com/robotframework/robotframework/blob/master/atest/robot/output/listener_interface/listener_v3.robot
https://github.com/robotframework/robotframework/blob/master/atest/testdata/output/listeners/v3.py

I'm not particularly happy with the attribute name `data` in listener v3 method documentation but other options aren't that good either. It was initially `model`, but that is somewhat confusing because it is a *data* model and the other argument is *result* model. It cannot really be `suite/test` either, because the other argument also represent suite or test. Notice that in the examples I actually used `suite/test` with that model object that was actually used.

**Comment by pekkaklarck** [comment]
I gave a new look at implementing `start_keyword` and `end_keyword` listener methods and my conclusion is that they needs to wait until RF 3.1 or later. It will require quite a bit refactoring, but I'm sure it would drive the internal design to better direction and would be worth it.

I consider this issue now done in RF 3.0 context. Enhancements to docs are obviously still possible.

**Comment by aaltat** [comment]
I did look the acceptance tests, the examples in the documentation are clearer and better. Writing separate issues about API and listener documentation is good idea.

**Comment by pekkaklarck** [comment]
Added #2279 about enhancing API docs. I try to get that done still before RF 3.0 final.

**Comment by pekkaklarck** [comment]
API docs have now been enhanced. Comments would be appreciated. See #2279 for details.

**Comment by pekkaklarck** [comment]
Adding `start/end_keyword` methods turned out to be even bigger task than anticipated, but it was finally done in RF 7.0 (#3296). After that the listener version 3 still lacks methods related to imports, but they are planned to be added in RF 7.1 (#5008).
